[SQL] CONCAT_WS for BINARY type fails with compilation error

The following SQL:
```
CREATE TABLE illegal_tbl(
bin BINARY);

CREATE MATERIALIZED VIEW v AS SELECT
CONCAT_WS('@', bin, 55) AS bin
FROM illegal_tbl;
```

With example input:
**"bin": [11, 22, 32]**

Fails with the following compilation error:
```
 error[E0425]: cannot find function `cast_to_sN_bytesN` in this scope
    --> crates/feldera_pipe_operator_dcf1ef152ecccf47/src/lib.rs:74:29
     |
74   |                 unwrap_cast(cast_to_sN_bytesN(p1.0.as_ref().cloned(),-1,false))),
     |                             ^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `cast_to_VN_bytesN`
     |
    ::: /Users/rivuadhikari/feldera/crates/sqllib/src/casts.rs:3629:1
     |
3629 | cast_to_variant!(bytes, ByteArray, Binary); // The other direction takes extra arguments
     | ------------------------------------------ similarly named function `cast_to_VN_bytesN` defined here
```

--- a/crates/sqllib/src/casts.rs
+++ b/crates/sqllib/src/casts.rs
@@ -4,7 +4,7 @@
 
 use crate::{
     array::Array,
-    binary::ByteArray,
+    binary::{to_hex_, ByteArray},
     dec::{is_error, DecimalContext, LargeDecimal, SqlDecimal},
     error::{r2o, SqlResult, SqlRuntimeError},
     geopoint::*,
@@ -1600,6 +1600,12 @@ pub fn cast_to_s_Uuid(value: Uuid, size: i32, fixed: bool) -> SqlResult<SqlStrin
     limit_or_size_string(&result, size, fixed)
 }
 
+#[doc(hidden)]
+pub fn cast_to_s_bytes(value: ByteArray, size: i32, fixed: bool) -> SqlResult<SqlString> {
+    let result = to_hex_(value);
+    limit_or_size_string(result.str(), size, fixed)
+}
+
 cast_to_string!(b, bool);
 cast_to_string!(SqlDecimal, SqlDecimal);
 cast_to_string!(f, F32);
@@ -1618,6 +1624,7 @@ cast_to_string!(u64, u64);
 cast_to_string!(Timestamp, Timestamp);
 cast_to_string!(Time, Time);
 cast_to_string!(Date, Date);
+cast_to_string!(bytes, ByteArray);
 cast_to_string!(LongInterval_MONTHS, LongInterval);
 cast_to_string!(LongInterval_YEARS, LongInterval);
 cast_to_string!(LongInterval_YEARS_TO_MONTHS, LongInterval);
@@ -3164,10 +3171,26 @@ cast_function!(Timestamp, Timestamp, Timestamp, Timestamp);
 #[doc(hidden)]
 #[inline]
 pub fn cast_to_Timestamp_i64(value: i64) -> SqlResult<Timestamp> {
-    Ok(Timestamp::new(value))
+    // Calcite ignores sub-second units
+    Ok(Timestamp::new((value / 1000) * 1000))
 }
 
 cast_function!(Timestamp, Timestamp, i64, i64);
+
+#[doc(hidden)]
+#[inline]
+pub fn cast_to_Timestamp_SqlDecimal(value: SqlDecimal) -> SqlResult<Timestamp> {
+    match TryInto::<i64>::try_into(value) {
+        // Calcite ignores sub-second units
+        Ok(value) => Ok(Timestamp::new((value / 1000) * 1000)),
+        Err(e) => Err(SqlRuntimeError::from_string(format!(
+            "Error converting {value} to TIMESTAMP: {}",
+            e
+        ))),
+    }
+}
+
+cast_function!(Timestamp, Timestamp, SqlDecimal, SqlDecimal);
 
 #[doc(hidden)]
 #[inline]

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresTimestampTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/postgres/PostgresTimestampTests.java
@@ -570,7 +570,7 @@ public class PostgresTimestampTests extends SqlIoTest {
                 SELECT FLOOR(timestamp '2004-02-29 15:44:17.71393' TO WEEK) AS week_trunc;
                         week_trunc
                 --------------------------
-                 Mon Feb 29 00:00:00 2004""");
+                 Sun Feb 29 00:00:00 2004""");
     }
 
     // DATE_BIN not supported by Calcite.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
@@ -247,7 +247,7 @@ public class CastTests extends SqlIoTest {
                 SELECT CAST(CAST(T.COL2 / 10 AS TIMESTAMP) AS DOUBLE) FROM T;
                  i
                 ---
-                 1
+                 0
                 (1 row)""");
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
@@ -418,4 +418,11 @@ public class Regression1Tests extends SqlIoTest {
             }
         });
     }
+
+    @Test
+    public void issue4255() {
+        this.getCCS("""
+                CREATE TABLE illegal_tbl(bin BINARY);
+                CREATE VIEW v AS SELECT CONCAT_WS('@', bin, 55) AS bin FROM illegal_tbl;""");
+    }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -419,6 +419,29 @@ public class RegressionTests extends SqlIoTest {
     }
 
     @Test
+    public void castBinaryToString() {
+        var ccs = this.getCCS("""
+                CREATE TABLE T(x BINARY(2));
+                CREATE VIEW V AS SELECT CAST(x AS VARCHAR), CAST(x'AB01' AS VARCHAR) FROM T;""");
+        ccs.step("INSERT INTO T VALUES(x'AB01')", """
+                    x | y    | weight
+                ----------------------
+                 ab01 | ab01 | 1""");
+    }
+
+    @Test
+    public void castDecimalToTimestamp() {
+        var ccs = this.getCCS("""
+                CREATE TABLE T(x DECIMAL(10, 2), y INT);
+                CREATE VIEW V AS SELECT CAST(x AS TIMESTAMP), CAST(10.20 AS TIMESTAMP),
+                CAST(10 AS TIMESTAMP), cast(y AS TIMESTAMP) FROM T;""");
+        ccs.step("INSERT INTO T VALUES(10.20, 10)", """
+                 x                   | decimal             | int                 | y                  | weight
+                -----------------------------------------------------------------------------------------------
+                 1970-01-01 00:00:00 | 1970-01-01 00:00:00 | 1970-01-01 00:00:00 | 1970-01-01 00:00:00 |1""");
+    }
+
+    @Test
     public void issue3073() {
         this.compileRustTestCase("""
                 CREATE MATERIALIZED VIEW v10_optimized AS

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/TableParser.java
@@ -47,27 +47,44 @@ import java.nio.charset.StandardCharsets;
 import java.sql.ResultSet;
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
-import java.text.ParseException;
-import java.text.SimpleDateFormat;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.time.format.DateTimeParseException;
+import java.time.temporal.ChronoField;
 import java.util.ArrayList;
-import java.util.Date;
 import java.util.List;
+import java.util.Locale;
 import java.util.UUID;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 /** Helper class for tests which is used to parse the expected output from queries */
 public class TableParser {
-    // Many of these are postgres specific
-    static final SimpleDateFormat[] TIMESTAMP_INPUT_FORMAT = {
-            new SimpleDateFormat("EEE MMM d HH:mm:ss.SSS yyyy"),
-            new SimpleDateFormat("EEE MMM d HH:mm:ss yyyy G"),
-            new SimpleDateFormat("EEE MMM d HH:mm:ss yyyy"),
-            new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
+    // Append second fractions
+    static DateTimeFormatterBuilder sf(String pattern) {
+        return new DateTimeFormatterBuilder()
+                .appendPattern(pattern)
+                .appendLiteral('.')
+                .appendFraction(ChronoField.MILLI_OF_SECOND, 1, 3, false);
+    }
+
+    // The first few of these are postgres specific
+    static final DateTimeFormatter[] TIMESTAMP_INPUT_FORMAT = {
+            sf("EEE MMM d HH:mm:ss")
+                    .appendPattern(" yyyy")
+                    .toFormatter(Locale.ENGLISH),
+            DateTimeFormatter.ofPattern("EEE MMM d HH:mm:ss yyyy G", Locale.ENGLISH),
+            DateTimeFormatter.ofPattern("EEE MMM d HH:mm:ss yyyy", Locale.ENGLISH),
+            sf("yyyy-MM-dd HH:mm:ss")
+                    .toFormatter(Locale.ENGLISH),
+            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.ENGLISH)
     };
-    static final SimpleDateFormat TIMESTAMP_OUTPUT_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.S");
-    static final SimpleDateFormat DATE_INPUT_FORMAT = new SimpleDateFormat("MM-dd-yyyy");
-    static final SimpleDateFormat DATE_OUTPUT_FORMAT = new SimpleDateFormat("yyyy-MM-dd");
+
+    static final DateTimeFormatter TIMESTAMP_OUTPUT_FORMAT = sf("yyyy-MM-dd HH:mm:ss").toFormatter(Locale.ENGLISH);
+    static final DateTimeFormatter DATE_INPUT_FORMAT = DateTimeFormatter.ofPattern("MM-dd-yyyy", Locale.ENGLISH);
+    static final DateTimeFormatter DATE_OUTPUT_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd", Locale.ENGLISH);
     static final Pattern YEAR = Pattern.compile("^(\\d+) years?(.*)");
     static final Pattern MONTHS = Pattern.compile("^\\s*(\\d+) months?(.*)");
     static final Pattern MINUS = Pattern.compile("^-\\s*(.*)");
@@ -78,7 +95,7 @@ public class TableParser {
     static final Pattern HMS = Pattern.compile("\\s*([0-9][0-9]:[0-9][0-9]:[0-9][0-9])(\\.[0-9]*)?(.*)");
     static final Pattern AGO = Pattern.compile("\\s*ago(.*)");
     @Nullable
-    static Date ZERO = null;
+    static LocalDateTime ZERO = null;
 
     /** Convert a timestamp from a format like Sat Feb 16 17:32:01 1996 to
      * a format like 1996-02-16 17:32:01 */
@@ -87,19 +104,19 @@ public class TableParser {
             return DBSPLiteral.none(type);
         try {
             if (ZERO == null)
-                ZERO = new SimpleDateFormat("yyyy-MM-dd").parse("0000-01-01");
-        } catch (ParseException e) {
+                ZERO = LocalDateTime.parse("0001-01-01 00:00:00", DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
+        } catch (DateTimeParseException e) {
             throw new RuntimeException(e);
         }
-        for (SimpleDateFormat input: TIMESTAMP_INPUT_FORMAT) {
+        for (DateTimeFormatter input: TIMESTAMP_INPUT_FORMAT) {
             String out;
             try {
                 // Calcite problems: does not support negative years, or fractional seconds ending in 0
-                Date converted = input.parse(timestamp);
+                LocalDateTime converted = LocalDateTime.parse(timestamp, input);
                 out = TIMESTAMP_OUTPUT_FORMAT.format(converted);
-                if (converted.before(ZERO))
+                if (converted.isBefore(ZERO))
                     out = "-" + out;
-            } catch (ParseException ignored) {
+            } catch (DateTimeParseException ignored) {
                 continue;
             }
             return new DBSPTimestampLiteral(out, type.mayBeNull);
@@ -115,17 +132,17 @@ public class TableParser {
         try {
             if (date.length() != 10)
                 throw new RuntimeException("Unexpected date " + date);
-            SimpleDateFormat inputFormat;
+            DateTimeFormatter inputFormat;
             if (date.charAt(2) == '-' && date.charAt(5) == '-')
                 inputFormat = DATE_INPUT_FORMAT;
             else if (date.charAt(4) == '-' && date.charAt(7) == '-')
                 inputFormat = DATE_OUTPUT_FORMAT;
             else
                 throw new RuntimeException("Unexpected date " + date);
-            Date converted = inputFormat.parse(date);
+            LocalDate converted = LocalDate.parse(date, inputFormat);
             String out = DATE_OUTPUT_FORMAT.format(converted);
             return new DBSPDateLiteral(out, type.mayBeNull);
-        } catch (ParseException ex) {
+        } catch (DateTimeParseException ex) {
             throw new RuntimeException("Could not parse " + date);
         }
     }

